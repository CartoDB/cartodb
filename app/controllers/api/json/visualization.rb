# encoding: utf-8
require 'sinatra/base'
require 'json'
require_relative '../../../models/visualization/member'

module CartoDB
  module Visualization
    class API < Sinatra::Base
      get '/api/v1/visualizations/:id' do
        begin
          member    = Member.new(id: params.fetch('id')).fetch
          response  = member.attributes.to_json
          [200, response]
        rescue KeyError
          [404]
        end
      end # get /api/v1/visualizations/:id
      
      post '/api/v1/visualizations' do
        member    = Member.new(payload).store
        response  = member.attributes.to_json
        [201, response]
      end # post /api/visualizations

      put '/api/v1/visualizations/:id' do
        begin
          member            = Member.new(id: params.fetch('id')).fetch
          member.attributes = payload
          member.store
          [200, member.attributes.to_json]
        rescue KeyError
          [404]
        end
      end # put /api/v1/visualizations/:id

      delete '/api/v1/visualizations/:id' do
        Member.new(id: params.fetch('id')).delete
        [204]
      end # delete '/api/v1/visualizations/:id

      helpers do
        def payload
          JSON.parse(request.body.read.to_s || String.new)
        end #payload
      end 
    end # Visualization
  end # API
end # CartoDB

