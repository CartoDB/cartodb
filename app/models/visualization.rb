# encoding: utf-8
require_relative '../../services/data-repository/repository'

module CartoDB
  module Visualization
    def self.default_repository=(repository)
      @repository = repository
    end # self.default_repository

    def self.default_repository
      @repository ||= DataRepository::Repository.new
    end #self.default_repository
  end # Visualization
end # CartoDB

