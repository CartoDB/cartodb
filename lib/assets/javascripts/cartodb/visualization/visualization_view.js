
/**
* Single table item in dashboard list
*
* Usage example:
*
var li = new VisualizationView({
model: model*,
limitation: !this.options.user.get("private_tables")
});

* It needs a table model to run correctly.
*
*/
var VisualizationView = cdb.core.View.extend({

  tagName: 'li',

  events: {
    "click a.delete:not(.disabled)": "_confirmAndDelete"
  },

  initialize: function() {
    _.bindAll(this, "render", "deleting", "deleted", "_confirmAndDelete");

    _.defaults(this.options, this.default_options);

    this.template = cdb.templates.getTemplate('dashboard/views/visualization_list_item');

    this.retrigger('saving', this.model);
    this.retrigger('saved', this.model);

    this.model.bind('change', this.render);
    this.bind("clean", this._reClean, this);
  },


  render: function() {
    var self = this;
    this.cleanTooltips();
    this.$el.html(this.template(this.model.toJSON()));
    this.$el.addClass('tableRow border');
    return this;
  },

  clean: function() {
    this.trigger('clean');
    this.elder('clean');
  },

  /**
  * Show delete confirmation after decides delete a visualization
  */
  _confirmAndDelete: function(ev) {
    var self = this;
    ev && (ev.preventDefault());

    this.delete_dialog = new cdb.admin.DeleteDialog({
      model: this.model,
      title: "Delete this visualization",
      ok_title: "Delete this visualization",
      content: 'You are about to delete this visualization.',
      config: this.options.config
    });

    $("body").append(this.delete_dialog.render().el);
    this.delete_dialog.open();

    this.delete_dialog.wait()
    .done(this.deleteVisualization.bind(this))
    .fail(this.cleanDeleteDialog.bind(this));
  },

  /**
  * Hides the content and show a notification saying the visualization is being deleted
  * @triggers deleting
  * @return undefined
  */
  deleting: function() {
    this.cleanTooltips();
    this.$el.addClass('disabled');
    this.$('a').addClass('disabled');
  },

  /**
  * Close the "deleting" notification and warns the user that the visualization has been deleted
  * @triggers deleted
  * @return undefined
  */
  deleted: function() {
    this.cleanTooltips();
    this.$el.html('');

    var notificationTpl =
    '<p class="dark">Your visualization (' + this.model.get("name") + ') has been deleted</p>' +
      '<a class="smaller close" href="#close">x</a>';
    var $container = $('<li class="flash"></li>');
    this.$el.after($container);
    this.notification = new cdb.ui.common.Notification({
      el: $container,
      timeout:3000,
      template: notificationTpl,
      hideMethod: 'fadeOut',
      showMethod: 'fadeIn'
    });

    this.notification.open();
    this.$el.remove();
    this.clean();
  },

  cleanDeleteDialog: function() {
    this.delete_dialog.clean();
  },

  deleteVisualization: function() {
    var self = this;
    this.deleting();
    this.model.destroy({wait: true})
    .done(this.deleted);
  },


  /**
  * Destroy droppable funcionality when el is being cleaned
  */
  _reClean: function() {
    this.$el.droppable("destroy");
  }
});


