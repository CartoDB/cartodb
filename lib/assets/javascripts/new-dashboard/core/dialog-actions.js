// Dialogs
import ChangePrivacy from 'new-dashboard/components/Backbone/Dialogs/ChangePrivacy';
import DeleteDialog from 'new-dashboard/components/Backbone/Dialogs/DeleteDialog';
import Dialog from 'new-dashboard/components/Backbone/Dialog.vue';
import DuplicateMap from 'new-dashboard/components/Backbone/Dialogs/DuplicateMap';
import LockMap from 'new-dashboard/components/Backbone/Dialogs/LockMap';
import MapMetadata from 'new-dashboard/components/Backbone/Dialogs/MapMetadata';
import Share from 'new-dashboard/components/Backbone/Dialogs/Share';

export function editMapMetadata (visualization) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <MapMetadata :visualization="visualization" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, MapMetadata }
    },
    { visualization }
  );
}

export function changePrivacy (visualization) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <ChangePrivacy :visualization="visualization" @close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, ChangePrivacy }
    },
    { visualization }
  );
}

export function shareVisualization (visualization) {
  this.showModal(
    {
      template: `
      <Dialog v-on:close="$emit('close')">
        <Share :visualization="visualization" v-on:close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, Share }
    },
    { visualization }
  );
}

export function duplicateMap (visualization) {
  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DuplicateMap :visualization="visualization" @close="$emit('close')"/>
      </Dialog>`,
      props: ['visualization'],
      components: { Dialog, DuplicateMap }
    },
    { visualization }
  );
}

export function changeLockState (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockMap
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          @deselectAll="deselectAll" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, LockMap },
      methods: {
        deselectAll: function () {
          return actionHandlers.deselectAll();
        }
      }
    }, dialogProps
  );
}

export function changeVisualizationsLockState (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <LockMap
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          @deselectAll="deselectAll" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, LockMap },
      methods: {
        deselectAll: function () {
          return actionHandlers.deselectAll();
        }
      }
    }, dialogProps
  );
}

export function deleteVisualization (visualization, contentType, actionHandlers) {
  const dialogProps = { visualization, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualization="visualization"
          :contentType="contentType"
          @close="$emit('close')"
          @deselectAll="deselectAll" />
      </Dialog>`,
      props: ['visualization', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        deselectAll: function () {
          return actionHandlers.deselectAll();
        }
      }
    }, dialogProps
  );
}

export function deleteVisualizations (visualizations, contentType, actionHandlers) {
  const dialogProps = { visualizations, contentType };

  this.showModal(
    {
      template: `
      <Dialog @close="$emit('close')">
        <DeleteDialog
          :visualizations="visualizations"
          :contentType="contentType"
          @close="$emit('close')"
          @deselectAll="deselectAll" />
      </Dialog>`,
      props: ['visualizations', 'contentType'],
      components: { Dialog, DeleteDialog },
      methods: {
        deselectAll: function () {
          return actionHandlers.deselectAll();
        }
      }
    }, dialogProps
  );
}
