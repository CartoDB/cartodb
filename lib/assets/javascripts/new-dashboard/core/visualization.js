import PermissionModel from 'dashboard/data/permission-model';

export function isShared (visualizationData, models) {
  return models.user.get('id') !== visualizationData.permission.owner.id;
}

export function getURL (visualizationData, models) {
  const loggedUserBackboneModel = models.user;
  const loggedUserId = loggedUserBackboneModel.id || loggedUserBackboneModel.get('id');
  const permissionModel = new PermissionModel(
    visualizationData.permission, { configModel: models.config }
  );

  let id = visualizationData.id;
  let userUrl = permissionModel.owner.viewUrl();

  if (loggedUserBackboneModel &&
      loggedUserId !== permissionModel.owner.get('id') &&
      permissionModel.hasAccess(loggedUserBackboneModel)) {
    userUrl = loggedUserBackboneModel.viewUrl();
    id = `${permissionModel.owner.get('username')}.${id}`;
  }

  return `${userUrl.toString()}/viz/${id}`;
}

export function getThumbnailUrl (visualizationData, models, options) {
  const protocol = isHTTPS() ? 'https' : 'http';
  const cdnConfig = models.config.get('cdn_url');
  const username = models.user.get('username');
  const thumbnailOptions = {
    protocol,
    username,
    mapsApiResource: models.config.getMapsResourceName(username),
    tpl: _generateImageTemplate(visualizationData.id),
    width: options.width,
    height: options.height,
    authTokens: _generateAuthTokensParams(visualizationData.auth_tokens)
  };

  if (cdnConfig) {
    Object.assign(thumbnailOptions, { cdn: cdnConfig[protocol] });
  }

  return interpolateTemplateUrl(cdnConfig ? 'cdn' : 'regular', thumbnailOptions);
}

const interpolateTemplateUrl = function (type, options) {
  if (type === 'regular') {
    return `${options.protocol}://${options.mapsApiResource}/api/v1/map/static/named/${options.tpl}/${options.width}/${options.height}.png${options.authTokens}`;
  }

  if (type === 'cdn') {
    return `${options.protocol}://${options.cdn}/${options.username}/api/v1/map/static/named/${options.tpl}/${options.width}/${options.height}.png${options.authTokens}`;
  }
};

const _generateAuthTokensParams = function (authTokens) {
  if (authTokens && authTokens.length > 0) {
    return '?' + authTokens.map(t => `auth_token=${t}`).join('&');
  }

  return '';
};

const _generateImageTemplate = function (visualizationId) {
  return `tpl_${visualizationId.replace(/-/g, '_')}`;
};

const isHTTPS = function () {
  return location.protocol.indexOf('https') === 0;
};
