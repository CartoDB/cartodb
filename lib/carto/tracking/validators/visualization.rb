module Carto
  module Tracking
    module Validators
      module Visualization
        module Helpers
          def visualization
            @format.fetch_record!(:visualization)
          end
        end

        module Writable
          include Carto::Tracking::Validators::Visualization::Helpers

          def check_visualization_writable!
            unless visualization.writable_by?(@reporter)
              raise Carto::UnauthorizedError.new
            end
          end
        end

        module Readable
          include Carto::Tracking::Validators::Visualization::Helpers

          def check_visualization_readable!
            unless visualization.is_accesible_by_user?(@reporter)
              raise Carto::UnauthorizedError.new
            end
          end
        end
      end
    end
  end
end
