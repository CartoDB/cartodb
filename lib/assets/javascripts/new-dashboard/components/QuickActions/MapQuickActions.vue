<template>
  <QuickActions v-if="!isShared" :actions="actions[actionMode]" v-on="getEventListeners()" @open="openQuickactions" @close="closeQuickactions"></QuickActions>
</template>

<script>
import QuickActions from 'new-dashboard/components/QuickActions/QuickActions';
import * as DialogActions from 'new-dashboard/core/dialog-actions';
import * as Visualization from 'new-dashboard/core/visualization';

export default {
  name: 'MapQuickActions',
  components: {
    QuickActions
  },
  props: {
    map: Object
  },
  computed: {
    actions () {
      return {
        mine: [
          { name: this.$t('QuickActions.maps.editInfo'), event: 'editInfo', shouldShow: true },
          { name: this.$t('QuickActions.maps.manageTags'), event: 'manageTags', shouldShow: true },
          { name: this.$t('QuickActions.maps.changePrivacy'), event: 'changePrivacy', shouldShow: true },
          { name: this.$t('QuickActions.maps.share'), event: 'shareVisualization', shouldShow: this.isUserInsideOrganization },
          { name: this.$t('QuickActions.maps.duplicate'), event: 'duplicateMap', shouldShow: true },
          { name: this.$t('QuickActions.maps.lock'), event: 'lockMap', shouldShow: true },
          { name: this.$t('QuickActions.maps.delete'), event: 'deleteMap', isDestructive: true, shouldShow: true }
        ],
        locked: [
          { name: this.$t('QuickActions.maps.unlock'), event: 'unlockMap', shouldShow: true }
        ]
      };
    },
    actionMode () {
      return this.map.locked ? 'locked' : 'mine';
    },
    isShared () {
      return Visualization.isShared(this.$props.map, this.$cartoModels);
    },
    isUserInsideOrganization () {
      const userOrganization = this.$store.state.user.organization;
      return userOrganization && userOrganization.id;
    }
  },
  methods: {
    getEventListeners () {
      const events = this.actions[this.actionMode].map(action => action.event);

      return events.reduce(
        (eventListeners, action) => {
          eventListeners[action] = this[action].bind(this);
          return eventListeners;
        }, {}
      );
    },
    showModal (componentDefinition, componentPropsData) {
      this.$modal.show(
        componentDefinition,
        componentPropsData,
        { width: '100%', height: '100%' }
      );
    },
    openQuickactions () {
      this.$emit('open');
    },
    closeQuickactions () {
      this.$emit('close');
    },
    editInfo () {
      DialogActions.editMapMetadata.apply(this, [this.map]);
    },
    changePrivacy () {
      DialogActions.changePrivacy.apply(this, [this.map]);
    },
    manageTags () {
      DialogActions.editMapMetadata.apply(this, [this.map]);
    },
    duplicateMap () {
      DialogActions.duplicateMap.apply(this, [this.map]);
    },
    unlockMap () {
      DialogActions.changeLockState.apply(this, [this.map, 'maps']);
    },
    lockMap () {
      DialogActions.changeLockState.apply(this, [this.map, 'maps']);
    },
    deleteMap () {
      DialogActions.deleteVisualization.apply(this, [this.map, 'maps']);
    },
    shareVisualization () {
      DialogActions.shareVisualization.apply(this, [this.map]);
    }
  }
};
</script>
