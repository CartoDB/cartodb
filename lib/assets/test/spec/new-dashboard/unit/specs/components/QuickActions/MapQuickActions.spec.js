import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import MapQuickActions from 'new-dashboard/components/QuickActions/MapQuickActions';
import visualization from '../../fixtures/visualizations';
import * as DialogActions from '../__mocks__/dialog-actions';
import usersArray from '../../fixtures/users';

// Backbone Models
import ConfigModel from 'dashboard/data/config-model';
import UserModel from 'dashboard/data/user-model';

const localVue = createLocalVue();
localVue.use(Vuex);

const $t = key => key;

function configCartoModels (attributes = {}) {
  const user = new UserModel(attributes.user || usersArray[0]);
  const config = new ConfigModel(attributes.config || { maps_api_template: 'http://{user}.example.com' });
  return { user, config };
}

let $cartoModels = configCartoModels();

let store = new Vuex.Store({
  state: {
    user: {
      organization: { id: 'fake-organization-id' }
    }
  }
});

describe('MapQuickAction.vue', () => {
  it('should render correct contents', () => {
    const map = visualization.visualizations[0];
    const mapQuickAction = shallowMount(MapQuickActions, {
      store,
      localVue,
      propsData: { map },
      mocks: {
        $cartoModels,
        $t
      }
    });

    expect(mapQuickAction).toMatchSnapshot();
  });

  it('should not render the component if the map is shared', () => {
    $cartoModels = configCartoModels({ user: usersArray[1] });
    const map = visualization.visualizations[0];
    const mapQuickAction = shallowMount(MapQuickActions, {
      propsData: { map },
      mocks: {
        $cartoModels,
        $t
      }
    });

    expect(mapQuickAction).toMatchSnapshot();
  });

  describe('computed', () => {
    it('should return "locked" action name when a map is locked', () => {
      const map = {...visualization.visualizations[0]};
      map.locked = true;
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      const result = mapQuickAction.vm.actionMode;

      expect(result).toBe('locked');
    });

    it('should return "mine" action name when a map is not locked', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      const result = mapQuickAction.vm.actionMode;

      expect(result).toBe('mine');
    });
  });

  describe('methods', () => {
    it('should call the function to show modal', () => {
      const $modal = {
        show: jest.fn()
      };
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $modal,
          $t
        }
      });

      mapQuickAction.vm.showModal();

      expect($modal.show).toHaveBeenCalled();
    });

    it('should emit event open when calling openQuickactions method', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });
      expect(mapQuickAction.emitted('open')).toBeFalsy();

      mapQuickAction.vm.openQuickactions();

      expect(mapQuickAction.emitted('open').length).toBe(1);
    });

    it('should emit event close when calling closeQuickactions method', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });
      expect(mapQuickAction.emitted('close')).toBeFalsy();

      mapQuickAction.vm.closeQuickactions();

      expect(mapQuickAction.emitted('close').length).toBe(1);
    });

    it('should open edit metadata modal', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.editInfo();

      expect(DialogActions.editMapMetadata).toHaveBeenCalled();
    });

    it('should open change privacy modal', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.changePrivacy();

      expect(DialogActions.changePrivacy).toHaveBeenCalled();
    });

    it('should open manage tags modal', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.manageTags();

      expect(DialogActions.editMapMetadata).toHaveBeenCalled();
    });

    it('should launch share with colleagues action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.shareVisualization();

      expect(DialogActions.shareVisualization).toHaveBeenCalled();
    });

    it('should launch duplicate map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.duplicateMap();

      expect(DialogActions.duplicateMap).toHaveBeenCalled();
    });

    it('should launch unlock map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.unlockMap();

      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should launch lock map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.lockMap();

      expect(DialogActions.changeLockState).toHaveBeenCalled();
    });

    it('should launch delete map action', () => {
      const map = visualization.visualizations[0];
      const mapQuickAction = shallowMount(MapQuickActions, {
        propsData: { map },
        mocks: {
          $cartoModels,
          $t
        }
      });

      mapQuickAction.vm.deleteMap();

      expect(DialogActions.deleteVisualization).toHaveBeenCalled();
    });
  });
});
