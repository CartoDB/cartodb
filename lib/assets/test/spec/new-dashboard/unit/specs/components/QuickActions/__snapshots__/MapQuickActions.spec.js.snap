// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapQuickAction.vue should not render the component if the map is shared 1`] = ``;

exports[`MapQuickAction.vue should render correct contents 1`] = `<quickactions-stub actions="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"></quickactions-stub>`;
